elastic/kibana
# Gracefully handle decryption errors during ESO migrations

Resolves https://github.com/elastic/kibana/issues/101582

## Summary

* Updated the ESO migration function to take a `shouldMigrateIfDecryptionFails` flag. If this flag is set to `false` or undefined, decryption errors will be thrown (this is the previous behavior). If the flag is set to `true`, decryption errors will be caught, encrypted attributes will be stripped and the migration function will be applied to the stripped document.
* Note that if there are errors during the actual migration or errors during encryption, they will continue to be thrown.
* Updated the rules and connectors migration function to set `shouldMigrateIfDecryptionFails` to `true` and to log and throw any errors thrown during migration. While the original issue specifically mentioned decryption errors, the try/catch block in these migration functions would actually swallow **any** error during migration (decryption, migrating, encryption) and skip migrating the saved object. 

## To verify
* Make sure you use `yarn es snapshot --ssl --license trial -E path.data=../data` to run ES so that you can test the migrations.
* Make sure you have value for `xpack.encryptedSavedObjects.encryptionKey` set in your kibana.yml
* To make use of existing rule/connector migrations, switch to a 7.10 version of Kibana and create some rules and email and/or webhook connectors. Alternatively, you can define a custom migration for 7.15.0 and just switch back to running 7.14.0.
* Switch to this branch. Change the value for `xpack.encryptedSavedObjects.encryptionKey` in your kibana.yml. Start up Kibana.
* Verify that the migrations for rules and connectors all ran by checking the saved object.
  * 7.11 rules should have new `updatedAt` and `notifyWhen` fields
  * 7.11 email/webhook connectors should have new `hasAuth` field
  * 7.14 all connectors should have new `isMissingSecrets` field
  * or if you defined your own migration for testing, that should have run.
* See that your rules & connectors are erroring with decrypt failures. These should be fixable by disabling/renabling the rule and reentering secrets for the connector.
* See warning messages in the logs informing you that saved object decryption failed but migration proceeded anyway.

### Checklist

Delete any items that are not applicable to this PR.
